cips/cip-network-upgrade.md

```diff
@@ -0,0 +1,75 @@
+
+Clients may query the tally for each version as follows:
+
+```protobuf
+message QueryVersionTallyRequest { uint64 version = 1; }
+
+message QueryVersionTallyResponse {
+  uint64 voting_power = 1;
+  uint64 threshold_power = 2;
+  uint64 total_voting_power = 3;
+}
+```
+
+When `voting_power` is greater or equal to `theshold_power` , the network MAY upgrade. This is done through a “crank” transaction, `MsgTryUpgrade`, which can be submitted by any account. The account covers the gas required for the calculation. If the quorum is met, the chain will perform the necessary migrations at the end of processing that block. The proposer of the following height will include the new version in the block. As is currently, nodes will only vote for blocks that match the same network version as theirs.
+
+If the network agrees to move to a version that is not supported by the node, the node will gracefully shutdown.
+
+## Rationale
+
+When attempting a major upgrade, there is increased risk that the network halts. At least 2/3 in voting power of the network is needed to migrate and agree upon the next block. However this does not take into account the actions of byzantine validators. As an example, say there is 100 in total voting power. 66 have signalled v2 and 33 have yet to signal i.e. they are still signalling v1. It takes 1 byzantine voting power to pass the threshold, convincing the network to propose v2 blocks and then omitting their vote leaving the network failing to reach consensus until one of the 33 are able to upgrade. At the other end of the spectrum, increasing the necessary threshold means less voting power required to veto the upgrade. The middle point here is setting a quorum of 5/6ths which provides 1/6 byzantine fault tolerance to liveness and requiring at least 1/6th of the network to veto the upgrade.
+
+Validators are permitted to signal for the current network version as a means of cancelling their prior decision to upgrade. This is important in the case that new information arises that convinces the network that the next version is not ready.
+
+An on-chain tallying system was decided over an off-chain as it canonicalises the information which is important for coordination and adds accountability to validators. As a result of this decision, validators will need to pay fees to upgrade which is necessary to avoid spamming the chain.
+
+## Backwards Compatibility
+
+This feature modifies the functionality of the state machine in a breaking way as the state machine can now dictate version changes. This will require a major upgrade to implement (thus the protocol won’t come into affect until the following major upgrade).
+
+As the API is additive, there is no need to consider backwards compatibility for clients.
+
+## Security Considerations
```

Comment:
Is there a Reference Implementation section?

Reply:
yes you can. I just meant having the section here, even if the implementation isn't ready. can just add to it "TBA" or something. so far, there's no Implementation section

Reply:
Done